### PR TITLE
e2e storage: test labels

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Adding tags with the new framework helper functions also adds them as Ginkgo labels. Labels enable much more flexible [filtering of specs](https://onsi.github.io/ginkgo/#filtering-specs). The goal is to register all tests this way, with no or only minimal changes to the full test names, so existing regular expressions should continue to work.

There's one exception: a space now gets inserted around tags, which wasn't always the case before. If a regular expression in a Prow job is sensitive to white space, then it will not match anymore.

#### Special notes for your reviewer:

This PR starts that work by converting the E2E storage framework, which is the tricky part. The [rest of the conversion](https://github.com/kubernetes/kubernetes/pull/121394) is just search/replace.  Because tags need to be added via function calls, tags embedded inside the driver definition YAML or JSON file no longer works. Some in-tree drivers use this to mark all of their tests as "slow" or as "depends on Windows" - this still works.

It's unclear if users need this. If they do, support could be added by allowing tags as a string slice and then converting those string entries to function calls. But that is complicated and not done yet.

#### Does this PR introduce a user-facing change?

```release-note
E2E storage tests: setting test tags like `[Slow]` via the DriverInfo.FeatureTag field is no longer supported.
```

/sig storage

### CODE REVIEW

**Comment by pohly** [comment]
The first commit removes a dot at the end after `]` to minimize changes during follow-up conversions. This obviously changes test names, but shouldn't matter.

The second commit makes more changes. Here's a comparison between the `go test -v ./test/e2e ./test/e2e_node ./test/e2e_kubeadm -args --list-tests` output before the last commit and after it. The diff comes from `diff <(sed -e 's/ *\(\.*[/a-z0-9_-]*.go\):[0-9]*: /\1: /' before) <(sed -e 's/ *\(\.*[/a-z0-9_-]*.go\):[0-9]*: /\1: /'  after)` which strips the source code location prefix.

Spaces get inserted, for example:
```patch
1127,1129c1127,1129
< storage/testsuites/disruptive.go: [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (block volmode)] disruptive[Disruptive][LinuxOnly] Should test that pv used in a pod that is deleted while the kubelet is down cleans up when the kubelet returns.
< storage/testsuites/disruptive.go: [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (block volmode)] disruptive[Disruptive][LinuxOnly] Should test that pv used in a pod that is force deleted while the kubelet is down cleans up when the kubelet returns.
< storage/testsuites/disruptive.go: [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (block volmode)] disruptive[Disruptive][LinuxOnly] Should test that pv written before kubelet restart is readable after restart.
---
> storage/testsuites/disruptive.go: [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (block volmode)] disruptive [Disruptive] [LinuxOnly] Should test that pv used in a pod that is deleted while the kubelet is down cleans up when the kubelet returns.
> storage/testsuites/disruptive.go: [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (block volmode)] disruptive [Disruptive] [LinuxOnly] Should test that pv used in a pod that is force deleted while the kubelet is down cleans up when the kubelet returns.
> storage/testsuites/disruptive.go: [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (block volmode)] disruptive [Disruptive] [LinuxOnly] Should test that pv written before kubelet restart is readable after restart.
```

When ignoring whitespace changes with `-w`, only this remains:
```patch
1285d1284
< storage/testsuites/snapshottable_stress.go: [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic Snapshot (delete policy)] snapshottable-stress[Feature:VolumeSnapshotDataSource] should support snapshotting of many volumes repeatedly [Slow] [Serial]
1288c1287
< storage/testsuites/snapshottable_stress.go: [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic Snapshot (retain policy)] snapshottable-stress[Feature:VolumeSnapshotDataSource] should support snapshotting of many volumes repeatedly [Slow] [Serial]
---
> storage/testsuites/snapshottable_stress.go: [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic Snapshot (delete policy)] snapshottable-stress [Feature:VolumeSnapshotDataSource] should support snapshotting of many volumes repeatedly [Slow] [Serial]
1290a1290
> storage/testsuites/snapshottable_stress.go: [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic Snapshot (retain policy)] snapshottable-stress [Feature:VolumeSnapshotDataSource] should support snapshotting of many volumes repeatedly [Slow] [Serial]
1570d1569
< storage/testsuites/snapshottable_stress.go: [sig-storage] CSI Volumes [Driver: pd.csi.storage.gke.io][Serial] [Testpattern: Dynamic Snapshot (delete policy)] snapshottable-stress[Feature:VolumeSnapshotDataSource] should support snapshotting of many volumes repeatedly [Slow] [Serial]
1573c1572
< storage/testsuites/snapshottable_stress.go: [sig-storage] CSI Volumes [Driver: pd.csi.storage.gke.io][Serial] [Testpattern: Dynamic Snapshot (retain policy)] snapshottable-stress[Feature:VolumeSnapshotDataSource] should support snapshotting of many volumes repeatedly [Slow] [Serial]
---
> storage/testsuites/snapshottable_stress.go: [sig-storage] CSI Volumes [Driver: pd.csi.storage.gke.io] [Serial] [Testpattern: Dynamic Snapshot (delete policy)] snapshottable-stress [Feature:VolumeSnapshotDataSource] should support snapshotting of many volumes repeatedly [Slow] [Serial]
1575a1575
> storage/testsuites/snapshottable_stress.go: [sig-storage] CSI Volumes [Driver: pd.csi.storage.gke.io] [Serial] [Testpattern: Dynamic Snapshot (retain policy)] snapshottable-stress [Feature:VolumeSnapshotDataSource] should support snapshotting of many volumes repeatedly [Slow] [Serial]
```

This also seems to be because of whitespaces - not sure why diff didn't ignore this. :shrug:

File: test/e2e/storage/testsuites/multivolume.go [hunk 1]
```diff
@@ -316,7 +318,7 @@ func (t *multiVolumeTestSuite) DefineTests(driver storageframework.TestDriver, p
 	// [        node1        ]
 	//   |                 |     <- same volume mode
 	// [volume1]   ->  [restored volume1 snapshot]
-	ginkgo.It("should concurrently access the volume and restored snapshot from pods on the same node [LinuxOnly][Feature:VolumeSnapshotDataSource][Feature:VolumeSourceXFS]", func(ctx context.Context) {
+	f.It("should concurrently access the volume and restored snapshot from pods on the same node [LinuxOnly]", feature.VolumeSnapshotDataSource, feature.VolumeSourceXFS, func(ctx context.Context) {
```
Comment (pohly) [inline test/e2e/storage/testsuites/multivolume.go:321]: This shows how tests will be registered going forward:
- `f.It` is a shorthand for `framework.It`, a wrapper around `ginkgo.It` which inserts the traditional inline text tags in addition to adding Ginkgo labels.
- `feature.VolumeSnapshotDataSource` is one of the registered feature tags from https://github.com/kubernetes/kubernetes/blob/master/test/e2e/feature/feature.go. No more typos and better visibility when new features get added... ideally those then should also come with documentation of what they mean. We don't have that for existing features.

File: test/e2e/storage/framework/testdriver.go [hunk 2]
```diff
@@ -219,7 +219,7 @@ type DriverInfo struct {
 	// plugin if it exists and is empty if this DriverInfo represents a CSI
 	// Driver
 	InTreePluginName string
-	FeatureTag       string // FeatureTag for the driver
+	TestTags         []interface{} // tags for the driver (e.g. framework.WithSlow())
```
Comment (pohly) [inline test/e2e/storage/framework/testdriver.go:222]: This field cannot be set via YAML or JSON, which affects external users of the e2e.test binary - see PR description.

**Comment by pohly** [comment]
Looks like there is one problem that I need to resolve:
> runtime error: comparing uncomparable type framework.label

**Comment by saad-ali** [comment]
Thanks for the improvement.

> There's one exception: a space now gets inserted around tags, which wasn't always the case before. If a regular expression in a Prow job is sensitive to white space, then it will not match anymore.

Hopefully that doesn't break anyone, seems fairly small, we can merge and find out.

/lgtm
/approve

**Pull request merged**